### ISSUE TITLE
Question about EMA update

### ISSUE THREAD

**Issue by totalhuang** [yifan123/flow_grpo#88]
I found that DPO and GRPO update EMA states in different places. DPO update EMA once sync_gradients as follows
```python
                    if accelerator.sync_gradients:
                        info = {k: torch.mean(torch.stack(v)) for k, v in info.items()}
                        info = accelerator.reduce(info, reduction="mean")
                        info.update({"epoch": epoch, "inner_epoch": inner_epoch})
                        if accelerator.is_main_process:
                            wandb.log(info, step=global_step)
                        info = defaultdict(list)
                        global_step += 1
                        if config.train.ema:
                            ema.step(transformer_trainable_parameters, global_step)
```
while GRPO update EMA when inner epoch finish as follows:
```python
                    if accelerator.sync_gradients:
                        info = {k: torch.mean(torch.stack(v)) for k, v in info.items()}
                        info = accelerator.reduce(info, reduction="mean")
                        info.update({"epoch": epoch, "inner_epoch": inner_epoch})
                        if accelerator.is_main_process:
                            wandb.log(info, step=global_step)
                        global_step += 1
                        info = defaultdict(list)
                if config.train.ema:
                    ema.step(transformer_trainable_parameters, global_step)
            # make sure we did an optimization step at the end of the inner epoch
```
Is it on purpose and why

**Comment by yifan123** [comment]
Thanks a lot for pointing out some issues in the code. It seems that I accidentally misaligned the indentation, which made it inconsistent with GRPO. However, because of the gradient accumulation, when it reaches `if accelerator.sync_gradients`, the loop should still exit, so the effect is the same. You can try changing it to match GRPO for consistency.
